src/spawn.c

```diff
@@ -0,0 +1,1121 @@
+ *      but the command line recreated by mscvrt may be wrong.
+ *    - GLib converts the argument vector to UNICODE. For non-UTF8 arguments,
+ *      the result is often "Invalid string in argument vector at %d: %s: Invalid
+ *      byte sequence in conversion input" (YMMV). Our tools (make, grep, gcc, ...)
+ *      are "ANSI", so converting to UNICODE and then back only causes problems.
+ *    - For various reasons, GLib uses an intermediate program to start children
+ *      (see gspawn-win32.c), the result being that children's children output
+ *      (such as make -> gcc) is lost.
+ *    - With non-blocking pipes, the g_io_add_watch() callbacks are not invoked,
+ *      while with blocking pipes, the child blocks on output larger than the
+ *      pipe buffer (4KB by default), and Geany blocks waiting for the child.
+ *      (g_io_channel_set_flags(G_IO_FLAG_NONBLOCK) does not work too, but that's
+ *      a small problem)
+ *
+ * Note that if no pipe I/O with the child is required, the g_spawn...() functions
+ * will generally work. They open the child in a normal window, while spawn_...()
+ * use a minimized-by-default window, more suitable for a tool.
+ *
+ * This module does not depend on Geany.
+ */
+
+#ifdef HAVE_CONFIG_H
+# include "config.h"
+#endif
+
+#include <errno.h>
+#include <fcntl.h>
+#include <string.h>
+
+#ifdef G_OS_WIN32
+# define _NO_W32_PSEUDO_MODIFIERS
+# include <windows.h>
+# include <errno.h>
+# include <io.h>
+#else  /* G_OS_WIN32 */
+# include <sys/types.h>
+# include <sys/wait.h>
+# include <signal.h>
+# include <unistd.h>
+# ifndef O_NONBLOCK
+# define O_NONBLOCK O_NDELAY
+# endif
+#endif  /* G_OS_WIN32 */
+
+#include "spawn.h"
+
+#ifndef _
+# define _
+#endif
+
+#ifdef G_OS_WIN32
+# define BLANKS " \t"
+#endif
+
+
+/**
+ *  Checks whether a command line is valid.
+ *
+ *  All OS:
+ *     - any leading spaces and tabs are ignored
+ *     - a command consisting solely of spaces and tabs is invalid
+ *  *nix:
+ *     - the standard shell quoting and escaping rules are used, see @c g_shell_parse_argv()
+ *     - as a consequence, an unqouted # at the start of an argument comments to the end of line
+ *  Windows:
+ *     - a quoted program name must be entirely inside the quotes ("C:\Foo\Bar".pdf is actually
+ *       started as "C:\Foo\Bar.exe")
+ *     - the standard Windows quoting rules are used: double quote is escaped with slash, and
+ *       any literal slashes before a double quote must be duplicated.
+ *
+ *  @param command_line the command line to check.
+ *  @param execute whether to check (using @c g_find_program_in_path() if the program specified
+ *         in the command line exists and is executable.
+ *  @param error return location for error.
+ *
+ *  @return @c TRUE on success, @c FALSE if an error was set.
+ **/
+gboolean spawn_check_command(const gchar *command_line, gboolean execute, GError **error)
+{
+	gchar *program;
+

```

Comment:
Maybe g_assert(), since we have control over the calling functions? Though personally I prefer the program to crash and display a stack frame in the debugger on obvious programming errors.